File: app/src/main/java/com/dialogapp/dialog/repository/PostsRepository.java
package com.dialogapp.dialog.repository;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.dialogapp.dialog.AppExecutors;
import com.dialogapp.dialog.api.ApiResponse;
import com.dialogapp.dialog.api.MicroblogService;
import com.dialogapp.dialog.db.MicroBlogDb;
import com.dialogapp.dialog.db.PostsDao;
import com.dialogapp.dialog.model.Item;
import com.dialogapp.dialog.model.MicroBlogResponse;
import com.dialogapp.dialog.util.Resource;

import java.util.List;
import java.util.concurrent.TimeUnit;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class PostsRepository {
    private final AppExecutors appExecutors;
    private final MicroblogService microblogService;
    private MicroBlogDb db;
    private final PostsDao postsDao;

    private RateLimiter<String> endpointRateLimit = new RateLimiter<>(2, TimeUnit.MINUTES);

    @Inject
    public PostsRepository(AppExecutors appExecutors, MicroBlogDb db, PostsDao postsDao, MicroblogService microblogService) {
        this.appExecutors = appExecutors;
        this.db = db;
        this.postsDao = postsDao;
        this.microblogService = microblogService;
    }

    public LiveData<Resource<List<Item>>> loadTimeline() {
        return new NetworkBoundResource<List<Item>, MicroBlogResponse>(appExecutors) {
            @Override
            protected boolean shouldFetch(@Nullable List<Item> dbData) {
                return dbData == null || dbData.isEmpty() || endpointRateLimit.shouldFetch(Endpoints.TIMELINE);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                return microblogService.getTimeLine(null);
            }

            @Override
            protected MicroBlogResponse processResponse(ApiResponse<MicroBlogResponse> response) {
                for (Item item : response.body.items) {
                    item.setEndpoint(Endpoints.TIMELINE);
                }
                return response.body;
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                db.beginTransaction();
                try {
                    db.postsDao().deletePosts(Endpoints.TIMELINE);
                    db.postsDao().insertPosts(response.items);
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
            }

            @NonNull
            @Override
            protected LiveData<List<Item>> loadFromDb() {
                return postsDao.loadEndpoint(Endpoints.TIMELINE);
            }

            @Override
            protected void onFetchFailed() {
                endpointRateLimit.reset(Endpoints.TIMELINE);
            }
        }.asLiveData();
    }

    public LiveData<Resource<List<Item>>> loadMentions() {
        return new NetworkBoundResource<List<Item>, MicroBlogResponse>(appExecutors) {
            @Override
            protected boolean shouldFetch(@Nullable List<Item> dbData) {
                return dbData == null || dbData.isEmpty() || endpointRateLimit.shouldFetch(Endpoints.MENTIONS);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                return microblogService.getMentions(null);
            }

            @Override
            protected MicroBlogResponse processResponse(ApiResponse<MicroBlogResponse> response) {
                for (Item item : response.body.items) {
                    item.setEndpoint(Endpoints.MENTIONS);
                }
                return response.body;
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                db.beginTransaction();
                try {
                    db.postsDao().deletePosts(Endpoints.MENTIONS);
                    db.postsDao().insertPosts(response.items);
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
            }

            @NonNull
            @Override
            protected LiveData<List<Item>> loadFromDb() {
                return postsDao.loadEndpoint(Endpoints.MENTIONS);
            }

            @Override
            protected void onFetchFailed() {
                endpointRateLimit.reset(Endpoints.MENTIONS);
            }
        }.asLiveData();
    }

    public LiveData<Resource<List<Item>>> loadFavorites() {
        return new NetworkBoundResource<List<Item>, MicroBlogResponse>(appExecutors) {
            @Override
            protected boolean shouldFetch(@Nullable List<Item> dbData) {
                return dbData == null || dbData.isEmpty() || endpointRateLimit.shouldFetch(Endpoints.FAVORITES);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                return microblogService.getFavorites();
            }

            @Override
            protected MicroBlogResponse processResponse(ApiResponse<MicroBlogResponse> response) {
                for (Item item : response.body.items) {
                    item.setEndpoint(Endpoints.FAVORITES);
                }
                return response.body;
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                postsDao.insertPosts(response.items);
            }

            @NonNull
            @Override
            protected LiveData<List<Item>> loadFromDb() {
                return postsDao.loadEndpoint(Endpoints.FAVORITES);
            }

            @Override
            protected void onFetchFailed() {
                endpointRateLimit.reset(Endpoints.FAVORITES);
            }
        }.asLiveData();
    }

    public LiveData<Resource<MicroBlogResponse>> loadPostsByUsername(String username) {
        return new NetworkBoundResource<MicroBlogResponse, MicroBlogResponse>(appExecutors) {
            @Override
            protected boolean shouldFetch(@Nullable MicroBlogResponse dbData) {
                return dbData == null || endpointRateLimit.shouldFetch(username);
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                return microblogService.getPostsByUsername(username);
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                postsDao.insertMicroblogData(response);
            }

            @NonNull
            @Override
            protected LiveData<MicroBlogResponse> loadFromDb() {
                return postsDao.loadMicroblogData(username);
            }

            @Override
            protected void onFetchFailed() {
                endpointRateLimit.reset(username);
            }
        }.asLiveData();
    }

    public LiveData<Resource<List<Item>>> loadConversation(String id) {
        return new NetworkBoundResource<List<Item>, MicroBlogResponse>(appExecutors) {
            List<Item> responseData;

            @Override
            protected boolean shouldFetch(@Nullable List<Item> dbData) {
                return true;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                return microblogService.getConversation(id);
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                responseData = response.items;
            }

            @NonNull
            @Override
            protected LiveData<List<Item>> loadFromDb() {
                return new LiveData<List<Item>>() {
                    @Override
                    protected void onActive() {
                        super.onActive();
                        setValue(responseData);
                    }
                };
            }
        }.asLiveData();
    }

    public LiveData<Resource<List<Item>>> loadDiscover(String topic) {
        return new NetworkBoundResource<List<Item>, MicroBlogResponse>(appExecutors) {
            @Override
            protected boolean shouldFetch(@Nullable List<Item> dbData) {
                return dbData == null || dbData.isEmpty() || endpointRateLimit.shouldFetch("discover");
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MicroBlogResponse>> createCall() {
                if (topic == null)
                    return microblogService.getFeaturedPosts();
                else
                    return microblogService.getFeaturedPosts(topic);
            }

            @Override
            protected MicroBlogResponse processResponse(ApiResponse<MicroBlogResponse> response) {
                for (Item item : response.body.items) {
                    item.setEndpoint(Endpoints.DISCOVER);
                }
                return response.body;
            }

            @Override
            protected void saveCallResult(@NonNull MicroBlogResponse response) {
                db.beginTransaction();
                try {
                    db.postsDao().deletePosts(Endpoints.DISCOVER);
                    db.postsDao().insertPosts(response.items);
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
            }

            @NonNull
            @Override
            protected LiveData<List<Item>> loadFromDb() {
                return postsDao.loadEndpoint(Endpoints.DISCOVER);
            }

            @Override
            protected void onFetchFailed() {
                endpointRateLimit.reset("discover");
            }
        }.asLiveData();
    }
}

File: app/src/main/java/com/dialogapp/dialog/repository/PostsRepository.kt
package com.dialogapp.dialog.repository

import androidx.lifecycle.LiveData
import com.dialogapp.dialog.AppExecutors
import com.dialogapp.dialog.api.ApiResponse
import com.dialogapp.dialog.api.MicroblogService
import com.dialogapp.dialog.db.InMemoryDb
import com.dialogapp.dialog.db.MicroBlogDb
import com.dialogapp.dialog.model.EndpointData
import com.dialogapp.dialog.model.FollowingAccount
import com.dialogapp.dialog.model.Post
import com.dialogapp.dialog.util.combineAndCompute
import com.dialogapp.dialog.vo.*
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton


@Singleton
class PostsRepository @Inject constructor(private val appExecutors: AppExecutors,
                                          private val db: MicroBlogDb,
                                          private val inMemoryDb: InMemoryDb,
                                          private val microblogService: MicroblogService) {

    private val endpointRateLimit = RateLimiter<String>(2, TimeUnit.MINUTES)

    fun loadDiscover(topic: String?, refresh: Boolean): LiveData<Resource<Listing<Post>>> {
        return object : NetworkBoundResource<Listing<Post>, MicroBlogResponse>(appExecutors) {
            override fun shouldFetch(data: Listing<Post>?): Boolean {
                return data?.postData.isNullOrEmpty() ||
                        (refresh && endpointRateLimit.shouldFetch(topic ?: DISCOVER))
            }

            override fun createCall(): LiveData<ApiResponse<MicroBlogResponse>> {
                return if (topic == null)
                    microblogService.getFeaturedPosts()
                else
                    microblogService.getFeaturedPosts(topic)
            }

            override fun saveCallResult(item: MicroBlogResponse) {
                inMemoryDb.runInTransaction {
                    val endpointData = EndpointData(endpoint = topic
                            ?: DISCOVER, microblog = item.microblog, author = item.author)
                    inMemoryDb.endpointData().insertEndpointData(endpointData)

                    item.posts.map {
                        it.belongsToEndpoint = topic ?: DISCOVER
                    }
                    inMemoryDb.posts().insertPosts(item.posts)
                }
            }

            override fun loadFromDb(): LiveData<Listing<Post>> {
                return inMemoryDb.posts().loadPostsByEndpointAsLiveData(endpoint = topic
                        ?: DISCOVER).combineAndCompute(inMemoryDb.endpointData().loadEndpointData(endpoint = topic
                        ?: DISCOVER)) { posts, endpointData ->
                    Listing(endpointData = endpointData, postData = posts)
                }
            }

            override fun onFetchFailed() {
                endpointRateLimit.reset(topic ?: DISCOVER)
            }
        }.asLiveData()
    }

    fun loadPostsByUsername(username: String, refresh: Boolean): LiveData<Resource<Listing<Post>>> {
        return object : NetworkBoundResource<Listing<Post>, MicroBlogResponse>(appExecutors) {
            override fun shouldFetch(data: Listing<Post>?): Boolean {
                return data?.postData.isNullOrEmpty() ||
                        (refresh && endpointRateLimit.shouldFetch(username))
            }

            override fun createCall(): LiveData<ApiResponse<MicroBlogResponse>> {
                return microblogService.getPostsByUsername(username)
            }

            override fun saveCallResult(item: MicroBlogResponse) {
                inMemoryDb.runInTransaction {
                    val endpointData = EndpointData(endpoint = username,
                            microblog = item.microblog, author = item.author)
                    inMemoryDb.endpointData().insertEndpointData(endpointData)

                    item.posts.map {
                        it.belongsToEndpoint = username
                    }
                    inMemoryDb.posts().insertPosts(item.posts)
                }
            }

            override fun loadFromDb(): LiveData<Listing<Post>> {
                return inMemoryDb.posts().loadPostsByEndpointAsLiveData(endpoint = username)
                        .combineAndCompute(inMemoryDb.endpointData().loadEndpointData(endpoint = username))
                        { posts, endpointData ->
                            Listing(endpointData = endpointData, postData = posts)
                        }
            }

            override fun onFetchFailed() {
                endpointRateLimit.reset(username)
            }
        }.asLiveData()
    }

    fun loadConversation(convId: String, refresh: Boolean): LiveData<Resource<Listing<Post>>> {
        return object : NetworkBoundResource<Listing<Post>, MicroBlogResponse>(appExecutors) {
            override fun shouldFetch(data: Listing<Post>?): Boolean {
                return data?.postData.isNullOrEmpty() ||
                        (refresh && endpointRateLimit.shouldFetch(convId))
            }

            override fun createCall(): LiveData<ApiResponse<MicroBlogResponse>> {
                return microblogService.getConversation(convId)
            }

            override fun saveCallResult(item: MicroBlogResponse) {
                inMemoryDb.runInTransaction {
                    val endpointData = EndpointData(endpoint = convId,
                            microblog = item.microblog, author = item.author)
                    inMemoryDb.endpointData().insertEndpointData(endpointData)

                    item.posts.map {
                        it.belongsToEndpoint = convId
                    }
                    inMemoryDb.posts().insertPosts(item.posts)
                }
            }

            override fun loadFromDb(): LiveData<Listing<Post>> {
                return inMemoryDb.posts().loadPostsByEndpointAsLiveData(endpoint = convId)
                        .combineAndCompute(inMemoryDb.endpointData().loadEndpointData(endpoint = convId))
                        { posts, endpointData ->
                            Listing(endpointData = endpointData, postData = posts)
                        }
            }

            override fun onFetchFailed() {
                endpointRateLimit.reset(convId)
            }
        }.asLiveData()
    }

    fun loadFavorites(refresh: Boolean): LiveData<Resource<Listing<Post>>> {
        return object : NetworkBoundResource<Listing<Post>, MicroBlogResponse>(appExecutors) {
            override fun shouldFetch(data: Listing<Post>?): Boolean {
                return data?.postData.isNullOrEmpty() ||
                        (refresh && endpointRateLimit.shouldFetch(FAVORITES))
            }

            override fun createCall(): LiveData<ApiResponse<MicroBlogResponse>> {
                return microblogService.getFavorites()
            }

            override fun saveCallResult(item: MicroBlogResponse) {
                inMemoryDb.runInTransaction {
                    val endpointData = EndpointData(endpoint = FAVORITES,
                            microblog = item.microblog, author = item.author)
                    inMemoryDb.endpointData().insertEndpointData(endpointData)

                    item.posts.map {
                        it.belongsToEndpoint = FAVORITES
                    }
                    inMemoryDb.posts().insertPosts(item.posts)
                }
            }

            override fun loadFromDb(): LiveData<Listing<Post>> {
                return inMemoryDb.posts().loadPostsByEndpointAsLiveData(endpoint = FAVORITES)
                        .combineAndCompute(inMemoryDb.endpointData().loadEndpointData(endpoint = FAVORITES))
                        { posts, endpointData ->
                            Listing(endpointData = endpointData, postData = posts)
                        }
            }

            override fun onFetchFailed() {
                endpointRateLimit.reset(FAVORITES)
            }
        }.asLiveData()
    }

    fun loadFollowing(username: String, isSelf: Boolean, refresh: Boolean): LiveData<Resource<List<FollowingAccount>>> {
        val endpoint = FOLLOWING.plus("_$username")
        return object : NetworkBoundResource<List<FollowingAccount>, List<FollowingAccount>>(appExecutors) {
            override fun shouldFetch(data: List<FollowingAccount>?): Boolean {
                return data == null || data.isEmpty() ||
                        (refresh && endpointRateLimit.shouldFetch(endpoint))
            }

            override fun createCall(): LiveData<ApiResponse<List<FollowingAccount>>> {
                return if (isSelf)
                    microblogService.getFollowingSelf(username)
                else
                    microblogService.getFollowing(username)
            }

            override fun saveCallResult(item: List<FollowingAccount>) {
                when {
                    isSelf -> {
                        db.runInTransaction {
                            val endpointData = EndpointData(endpoint = endpoint,
                                    microblog = null, author = null)
                            db.endpointData().insertEndpointData(endpointData)

                            item.map {
                                it.belongsToEndpoint = endpoint
                            }
                            db.followingData().insertFollowingAccounts(item)
                        }
                    }
                    else -> {
                        inMemoryDb.runInTransaction {
                            val endpointData = EndpointData(endpoint = endpoint,
                                    microblog = null, author = null)
                            inMemoryDb.endpointData().insertEndpointData(endpointData)

                            item.map {
                                it.belongsToEndpoint = endpoint
                            }
                            inMemoryDb.followingData().insertFollowingAccounts(item)
                        }
                    }
                }
            }

            override fun loadFromDb(): LiveData<List<FollowingAccount>> {
                return when {
                    isSelf -> db.followingData().loadFollowingAccounts(endpoint)
                    else -> inMemoryDb.followingData().loadFollowingAccounts(endpoint)
                }
            }

            override fun onFetchFailed() {
                endpointRateLimit.reset(endpoint)
            }
        }.asLiveData()
    }
}
